File: solution/python/normalizer/execution.py
import csv
import logging
import normalizer
import validator
from generated import normalized_log_pb2 as nlpb

class ExecutionNormalizer(normalizer.NormalizerInterface):

    # Constructor.
    def __init__(self, input="", binaryOutput="", jsonOutput=""):
        self.input = input
        self.binaryOutput = binaryOutput
        self.jsonOutput=jsonOutput

    # Return input file location.
    def getInput(self) -> str:
        return self.input
    
    # Return binary file location.
    def getBinaryOutput(self) -> str:
        return self.binaryOutput
    
    # Return JSON file location.
    def getJsonOutput(self) -> str:
        return self.jsonOutput
    
    # Normalize log source line-by-line.
    def normalize(self, line="") -> nlpb.NormalizedLog:
        # Initialise protobuf log structures
        log = nlpb.NormalizedLog()

        # Parse line using csv library due to presence of commas in data
        fields = list(csv.reader([line]))[0]

        if len(fields) != 9:
            logging.warning("invalid number of fields found; expect 9, found %d: %s",len(fields), line)
            return None
        
        # Parse and populate timestamp
        try:
            ts = validator.validateTimestamp(fields[0])
            log.execution_log.timestamp.CopyFrom(ts)
        except Exception as err:
            logging.warning("%s, skipping: %s", err, line)
            return None
        
        # Parse and populate filepath
        log.execution_log.filepath = fields[1].strip("\"")

        # Parse and populate command
        log.execution_log.command = fields[2].strip("\"")

        # Parse and populate uid
        try:
            uid = validator.validateInt(fields[3])
            log.execution_log.uid = uid
        except Exception as err:
            logging.warning("%s, skipping: %s", err, line)
            return None

        # Parse and populate pid
        try:
            pid = validator.validateInt(fields[4])
            log.execution_log.pid = pid
        except Exception as err:
            logging.warning("%s, skipping: %s", err, line)
            return None
        
        # Parse and populate ppid
        try:
            ppid = validator.validateInt(fields[5])
            log.execution_log.ppid = ppid
        except Exception as err:
            logging.warning("%s, skipping: %s", err, line)
            return None
        
        # Parse and populate cwd
        log.execution_log.cwd = fields[6].strip("\"")

        # Parse and populate hostname
        log.execution_log.hostname = fields[7].strip("\"")

        # Parse and populate platform
        try:
            log.execution_log.platform = validator.validatePlatform(fields[8])
        except Exception as err:
            logging.warning("%s, skipping: %s", err, line)
            return None
        
        return log

File: solution/go/normalizer/execution.go
package main

import (
	"log"
	"strings"

	nlpb "github.com/bearlyrunning/FindingTheNeedle/solution/go/generated/normalizedlogpb"
)

func (en *ExecutionNormalizer) getInput() string {
	return en.input
}

func (en *ExecutionNormalizer) getBinaryOutput() string {
	return en.binaryOutput
}

func (en *ExecutionNormalizer) getJsonOutput() string {
	return en.jsonOutput
}

func splitWithEscape(str, sep, esc string) []string {
	strs := strings.Split(str, sep)

	var escapedStrs []string
	concat := false
	concatStr := ""
	for _, s := range strs {
		// If current element ends with "\", enable concatnation.
		if strings.HasSuffix(s, esc) {
			if !concat {
				concat = true
			}
			concatStr += strings.ReplaceAll(s, "\\", ",")
		} else {
			concatStr += s
			escapedStrs = append(escapedStrs, concatStr)
			concat = false
			concatStr = ""
		}
	}
	return escapedStrs
}

func (en *ExecutionNormalizer) normalize(line string) *nlpb.NormalizedLog {
	fields := splitWithEscape(line, ",", "\\")

	// Validate fields.
	if len(fields) != 9 {
		log.Printf("invalid number of fields found; expect 9, found %d: %s\n", len(fields), line)
		return nil
	}
	timestamp, err := validateTimestamp(fields[0])
	if err != nil {
		log.Printf("%v, skipping: %s\n", err, line)
		return nil
	}
	uid, err := validateInt64(fields[3])
	if err != nil {
		log.Printf("%v, skipping: %s\n", err, line)
		return nil
	}
	pid, err := validateInt64(fields[4])
	if err != nil {
		log.Printf("%v, skipping: %s\n", err, line)
		return nil
	}
	ppid, err := validateInt64(fields[5])
	if err != nil {
		log.Printf("%v, skipping: %s\n", err, line)
		return nil
	}
	platform := validatePlatform(fields[8])
	return &nlpb.NormalizedLog{
		Msg: &nlpb.NormalizedLog_ExecutionLog{
			ExecutionLog: &nlpb.Execution{
				Timestamp: timestamp,
				Filepath:  strings.Trim(fields[1], "\""),
				Command:   strings.Trim(fields[2], "\""),
				Uid:       uid,
				Pid:       pid,
				Ppid:      ppid,
				Cwd:       strings.Trim(fields[6], "\""),
				Hostname:  strings.Trim(fields[7], "\""),
				Platform:  platform,
			},
		},
	}
}
